test/RegisterSpec.hs

```diff
@@ -2,40 +2,94 @@
 
 module RegisterSpec where
 
-import Register
-import Test.Hspec
-import Test.QuickCheck
 import Bit
+import Control.Monad
 import FixedVector
 import FixedVectorSpec ()
 import GHC.TypeLits
+import Register
+import Test.Hspec
+import Test.QuickCheck
 import Prelude hiding ((+))
-import Control.Monad
 
 notChangesEachBit :: KnownNat n => Register n -> Expectation
 notChangesEachBit r = zipWithM_ shouldNotBe (toList r) (toList $ Register.not r)
 
 notIsItsOwnInverse :: KnownNat n => Register n -> Expectation
 notIsItsOwnInverse r = Register.not (Register.not r) `shouldBe` r
 
+allZeros :: KnownNat n => Register n
 allZeros = fromList [O, O ..]
+
+allOnes :: KnownNat n => Register n
 allOnes = fromList [I, I ..]
 
 testAnd :: KnownNat n => Register n -> Expectation
 testAnd r = do
     Register.and r r `shouldBe` r
-    Register.and r (Register.not r) `shouldBe` fromList [O, O ..]
-    -- TODO(thies): and with allZeros and allOnes, implement other tests
+    Register.and r (Register.not r) `shouldBe` allZeros
+    Register.and r allZeros `shouldBe` allZeros
 
 testOr :: KnownNat n => Register n -> Expectation
 testOr r = do
     Register.or r r `shouldBe` r
-    Register.or r (Register.not r) `shouldBe` fromList [I, I ..]
+    Register.or r (Register.not r) `shouldBe` allOnes
+
+testOrNotZero :: KnownNat n => Register n -> Expectation
+testOrNotZero r
+    | r == allZeros = Register.or r allZeros `shouldBe` allZeros
+    | otherwise = Register.or r allZeros `shouldNotBe` allZeros
+
+testNand :: KnownNat n => Register n -> Register n -> Expectation
+testNand r1 r2 = Register.nand r1 r2 `shouldBe` (Register.not $ Register.and r1 r2)
+
+testNor :: KnownNat n => Register n -> Register n -> Expectation
+testNor r1 r2 = Register.nor r1 r2 `shouldBe` (Register.not $ Register.or r1 r2)
+
+testXor :: KnownNat n => Register n -> Expectation
+testXor r = do
+    Register.xor r (Register.not r) `shouldBe` allOnes
+    Register.xor r r `shouldBe` allZeros
+
+testXnor :: KnownNat n => Register n -> Register n -> Expectation
+testXnor r1 r2 = Register.xnor r1 r2 `shouldBe` (Register.not $ Register.xor r1 r2)
+
+impliesAllZeros :: KnownNat n => Register n -> Register n -> Expectation
+impliesAllZeros r1 r2
+    | r2 == allZeros = Register.implies allOnes r2 `shouldBe` allZeros
+    | otherwise = Register.implies r1 r2 `shouldNotBe` allZeros
+
+impliesAllOnes :: KnownNat n => Register n -> Expectation
+impliesAllOnes r = Register.implies r allOnes `shouldBe` allOnes
```

Comment:
I felt implication is a special case because the result is always I unless `implies I O`, so why not explicitly cover the only case where the result is allZeros and the only case we can predict it's allOnes?

Reply:
I don't think it is good form to write a property test (a statement that should hold for any given register) and then treat a special case of it (all zeroes) as a unit test. That is regardless of the function you are testing, so I don't think implication is a special case at all.

@joto314 can you see any nice properties of implication we can test?

Reply:
Implies A B is just (not A) or B, that could be tested. Alternatively you have things like (implies A A) is all ones and "implies A (not A) == not A" and "implies (not A) A == A".

But to be honest, for me the characterization of Implies is that it is false if and only if "I implies O": "Something false can not follow from something true".

Reply:
One thing to keep in mind is that we don’t want the test code to basically duplicate the implementation.

For example, this test:
```haskell
implies a b `shouldBe` (not a) b
```
doesn’t really test any properties of `implies`, it just suggests an (alternative) implementation for it. What’s worse, this implementation depends on both `not` and `or` working as intended.